File: packages/sardis-sdk-js/src/resources/wallets.ts
/**
 * Wallets Resource
 * 
 * Manages non-custodial wallets for agents.
 * Wallets never hold funds - they only sign transactions via MPC.
 */

import { BaseResource } from './base.js';
import type {
  Wallet,
  WalletBalance,
  CreateWalletInput,
  SetAddressInput,
  WalletTransferInput,
  WalletTransferResponse,
  RequestOptions,
} from '../types.js';

export class WalletsResource extends BaseResource {
  private _normalize(wallet: Wallet): Wallet {
    // Provide backwards-compatible alias: id := wallet_id
    if (!wallet.id) {
      (wallet as unknown as { id: string }).id = wallet.wallet_id;
    }
    return wallet;
  }

  /**
   * Create a new non-custodial wallet for an agent.
   *
   * @param input Wallet creation parameters
   * @param options Request options (signal, timeout)
   * @returns Created wallet
   */
  async create(input: CreateWalletInput, options?: RequestOptions): Promise<Wallet> {
    const wallet = await this._post<Wallet>('/api/v2/wallets', input, options);
    return this._normalize(wallet);
  }

  /**
   * Get a wallet by ID.
   *
   * @param walletId Wallet ID
   * @param options Request options (signal, timeout)
   * @returns Wallet details
   */
  async get(walletId: string, options?: RequestOptions): Promise<Wallet> {
    const wallet = await this._get<Wallet>(`/api/v2/wallets/${walletId}`, undefined, options);
    return this._normalize(wallet);
  }

  /**
   * List wallets.
   *
   * @param agentId Optional agent ID filter
   * @param limit Maximum number of wallets to return
   * @param options Request options (signal, timeout)
   * @returns List of wallets
   */
  async list(agentId?: string, limit: number = 100, options?: RequestOptions): Promise<Wallet[]> {
    const params: Record<string, unknown> = { limit };
    if (agentId) {
      params.agent_id = agentId;
    }
    const response = await this._get<{ wallets: Wallet[] } | Wallet[]>(
      '/api/v2/wallets',
      params,
      options
    );

    // Handle both array and object response formats
    if (Array.isArray(response)) {
      return response.map((w) => this._normalize(w));
    }
    return (response.wallets || []).map((w) => this._normalize(w));
  }

  /**
   * Get wallet balance from chain (non-custodial, read-only).
   *
   * @param walletId Wallet ID
   * @param chain Chain identifier (e.g., "base", "polygon")
   * @param token Token symbol (e.g., "USDC", "USDT")
   * @param options Request options (signal, timeout)
   * @returns Wallet balance from chain
   */
  async getBalance(
    walletId: string,
    chain: string = 'base',
    token: string = 'USDC',
    options?: RequestOptions
  ): Promise<WalletBalance> {
    return this._get<WalletBalance>(`/api/v2/wallets/${walletId}/balance`, {
      chain,
      token,
    }, options);
  }

  /**
   * Get all wallet addresses (chain -> address mapping).
   *
   * @param walletId Wallet ID
   * @param options Request options (signal, timeout)
   * @returns Dictionary mapping chain names to addresses
   */
  async getAddresses(walletId: string, options?: RequestOptions): Promise<Record<string, string>> {
    return this._get<Record<string, string>>(
      `/api/v2/wallets/${walletId}/addresses`,
      undefined,
      options
    );
  }

  /**
   * Set wallet address for a chain.
   *
   * @param walletId Wallet ID
   * @param input Address information
   * @param options Request options (signal, timeout)
   * @returns Updated wallet
   */
  async setAddress(
    walletId: string,
    input: SetAddressInput,
    options?: RequestOptions
  ): Promise<Wallet> {
    const wallet = await this._post<Wallet>(
      `/api/v2/wallets/${walletId}/addresses`,
      input,
      options
    );
    return this._normalize(wallet);
  }

  /**
   * Transfer stablecoins from a wallet (agent is sender).
   *
   * This endpoint is intended to be called by an agent process using an API key.
   * Sardis enforces policy + compliance and signs via the agent MPC wallet.
   */
  async transfer(
    walletId: string,
    input: WalletTransferInput,
    options?: RequestOptions
  ): Promise<WalletTransferResponse> {
    return this._post<WalletTransferResponse>(
      `/api/v2/wallets/${walletId}/transfer`,
      input,
      options
    );
  }

  async upgradeSmartAccount(
    walletId: string,
    input: {
      smart_account_address: string;
      entrypoint_address?: string;
      paymaster_enabled?: boolean;
      bundler_profile?: string;
    },
    options?: RequestOptions
  ): Promise<Wallet> {
    const wallet = await this._post<Wallet>(
      `/api/v2/wallets/${walletId}/upgrade-smart-account`,
      input,
      options
    );
    return this._normalize(wallet);
  }
}

File: packages/sardis-sdk-python/src/sardis_sdk/resources/wallets.py
"""
Wallets resource for Sardis SDK.

This module provides both async and sync interfaces for managing wallets.
Wallets hold funds and execute payments for agents.
"""
from __future__ import annotations

from decimal import Decimal
from typing import TYPE_CHECKING, Any, Dict, List, Optional, Union

from ..models import Wallet, WalletBalance, WalletCreate, WalletTransferResponse
from .base import AsyncBaseResource, SyncBaseResource

if TYPE_CHECKING:
    from ..client import AsyncSardisClient, SardisClient, TimeoutConfig


class AsyncWalletsResource(AsyncBaseResource):
    """Async resource for managing wallets.

    Wallets hold funds and execute payments. Each wallet:
    - Belongs to an Agent
    - Supports multiple tokens (USDC, USDT, etc.)
    - Has configurable spending limits
    - Is non-custodial (MPC-based)

    Example:
        ```python
        async with AsyncSardisClient(api_key="...") as client:
            # Create a wallet
            wallet = await client.wallets.create(
                agent_id="agent_123",
                currency="USDC",
            )

            # Get wallet balance
            balance = await client.wallets.get_balance("wallet_123")

            # List wallets for an agent
            wallets = await client.wallets.list(agent_id="agent_123")
        ```
    """

    async def create(
        self,
        agent_id: str,
        mpc_provider: str = "turnkey",
        currency: str = "USDC",
        chain: Optional[str] = None,
        limit_per_tx: Optional[Decimal] = None,
        limit_total: Optional[Decimal] = None,
        metadata: Optional[Dict[str, Any]] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Wallet:
        """Create a new non-custodial wallet for an agent.

        Args:
            agent_id: ID of the owner agent
            mpc_provider: MPC provider ("turnkey", "fireblocks", or "local")
            currency: Base currency code (default: USDC)
            chain: Optional chain identifier (e.g., "base_sepolia", "base")
            limit_per_tx: Per-transaction spending limit
            limit_total: Total spending limit
            metadata: Optional metadata dictionary
            timeout: Optional request timeout

        Returns:
            The created Wallet object
        """
        payload: Dict[str, Any] = {
            "agent_id": agent_id,
            "mpc_provider": mpc_provider,
            "currency": currency,
        }

        if chain is not None:
            payload["chain"] = chain
        if limit_per_tx is not None:
            payload["limit_per_tx"] = str(limit_per_tx)
        if limit_total is not None:
            payload["limit_total"] = str(limit_total)
        if metadata is not None:
            payload["metadata"] = metadata

        data = await self._post("wallets", payload, timeout=timeout)
        return Wallet.model_validate(data)

    async def get(
        self,
        wallet_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Wallet:
        """Get a wallet by ID.

        Args:
            wallet_id: The ID of the wallet to retrieve
            timeout: Optional request timeout

        Returns:
            The Wallet object
        """
        data = await self._get(f"wallets/{wallet_id}", timeout=timeout)
        return Wallet.model_validate(data)

    async def list(
        self,
        agent_id: Optional[str] = None,
        limit: int = 100,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> List[Wallet]:
        """List wallets.

        Args:
            agent_id: Filter by owner agent ID
            limit: Maximum number of wallets to return
            timeout: Optional request timeout

        Returns:
            List of Wallet objects
        """
        params: Dict[str, Any] = {"limit": limit}
        if agent_id:
            params["agent_id"] = agent_id

        data = await self._get("wallets", params=params, timeout=timeout)

        if isinstance(data, list):
            return [Wallet.model_validate(item) for item in data]
        return [Wallet.model_validate(item) for item in data.get("wallets", data.get("items", []))]

    async def get_balance(
        self,
        wallet_id: str,
        chain: str = "base",
        token: str = "USDC",
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> WalletBalance:
        """Get wallet balance from chain (non-custodial, read-only).

        Args:
            wallet_id: The wallet ID
            chain: Chain identifier (e.g., "base", "polygon")
            token: The token symbol (e.g., USDC, USDT)
            timeout: Optional request timeout

        Returns:
            WalletBalance object with balance from chain
        """
        params = {"chain": chain, "token": token}
        data = await self._get(f"wallets/{wallet_id}/balance", params=params, timeout=timeout)
        return WalletBalance.model_validate(data)

    async def get_addresses(
        self,
        wallet_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Dict[str, str]:
        """Get all wallet addresses (chain -> address mapping).

        Args:
            wallet_id: The wallet ID
            timeout: Optional request timeout

        Returns:
            Dictionary mapping chain names to addresses
        """
        return await self._get(f"wallets/{wallet_id}/addresses", timeout=timeout)

    async def set_address(
        self,
        wallet_id: str,
        chain: str,
        address: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Wallet:
        """Set wallet address for a chain.

        Args:
            wallet_id: The wallet ID
            chain: Chain identifier (e.g., "base", "polygon")
            address: Wallet address on the chain
            timeout: Optional request timeout

        Returns:
            Updated Wallet object
        """
        payload = {"chain": chain, "address": address}
        data = await self._post(f"wallets/{wallet_id}/addresses", payload, timeout=timeout)
        return Wallet.model_validate(data)

    async def update(
        self,
        wallet_id: str,
        limit_per_tx: Optional[Decimal] = None,
        limit_total: Optional[Decimal] = None,
        is_active: Optional[bool] = None,
        metadata: Optional[Dict[str, Any]] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Wallet:
        """Update a wallet.

        Args:
            wallet_id: The wallet ID
            limit_per_tx: New per-transaction spending limit
            limit_total: New total spending limit
            is_active: Enable/disable the wallet
            metadata: New metadata
            timeout: Optional request timeout

        Returns:
            Updated Wallet object
        """
        payload: Dict[str, Any] = {}
        if limit_per_tx is not None:
            payload["limit_per_tx"] = str(limit_per_tx)
        if limit_total is not None:
            payload["limit_total"] = str(limit_total)
        if is_active is not None:
            payload["is_active"] = is_active
        if metadata is not None:
            payload["metadata"] = metadata

        data = await self._patch(f"wallets/{wallet_id}", payload, timeout=timeout)
        return Wallet.model_validate(data)

    async def transfer(
        self,
        wallet_id: str,
        *,
        destination: str,
        amount: Decimal,
        token: str = "USDC",
        chain: str = "base_sepolia",
        domain: str = "localhost",
        memo: Optional[str] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> WalletTransferResponse:
        """Transfer stablecoins from a wallet (agent is sender)."""
        payload: Dict[str, Any] = {
            "destination": destination,
            "amount": str(amount),
            "token": token,
            "chain": chain,
            "domain": domain,
        }
        if memo is not None:
            payload["memo"] = memo
        data = await self._post(f"wallets/{wallet_id}/transfer", payload, timeout=timeout)
        return WalletTransferResponse.model_validate(data)


class WalletsResource(SyncBaseResource):
    """Sync resource for managing wallets.

    Wallets hold funds and execute payments. Each wallet:
    - Belongs to an Agent
    - Supports multiple tokens (USDC, USDT, etc.)
    - Has configurable spending limits
    - Is non-custodial (MPC-based)

    Example:
        ```python
        with SardisClient(api_key="...") as client:
            # Create a wallet
            wallet = client.wallets.create(
                agent_id="agent_123",
                currency="USDC",
            )

            # Get wallet balance
            balance = client.wallets.get_balance("wallet_123")

            # List wallets for an agent
            wallets = client.wallets.list(agent_id="agent_123")
        ```
    """

    def create(
        self,
        agent_id: str,
        mpc_provider: str = "turnkey",
        currency: str = "USDC",
        chain: Optional[str] = None,
        limit_per_tx: Optional[Decimal] = None,
        limit_total: Optional[Decimal] = None,
        metadata: Optional[Dict[str, Any]] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Wallet:
        """Create a new non-custodial wallet for an agent.

        Args:
            agent_id: ID of the owner agent
            mpc_provider: MPC provider ("turnkey", "fireblocks", or "local")
            currency: Base currency code (default: USDC)
            chain: Optional chain identifier (e.g., "base_sepolia", "base")
            limit_per_tx: Per-transaction spending limit
            limit_total: Total spending limit
            metadata: Optional metadata dictionary
            timeout: Optional request timeout

        Returns:
            The created Wallet object
        """
        payload: Dict[str, Any] = {
            "agent_id": agent_id,
            "mpc_provider": mpc_provider,
            "currency": currency,
        }

        if chain is not None:
            payload["chain"] = chain
        if limit_per_tx is not None:
            payload["limit_per_tx"] = str(limit_per_tx)
        if limit_total is not None:
            payload["limit_total"] = str(limit_total)
        if metadata is not None:
            payload["metadata"] = metadata

        data = self._post("wallets", payload, timeout=timeout)
        return Wallet.model_validate(data)

    def get(
        self,
        wallet_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Wallet:
        """Get a wallet by ID.

        Args:
            wallet_id: The ID of the wallet to retrieve
            timeout: Optional request timeout

        Returns:
            The Wallet object
        """
        data = self._get(f"wallets/{wallet_id}", timeout=timeout)
        return Wallet.model_validate(data)

    def list(
        self,
        agent_id: Optional[str] = None,
        limit: int = 100,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> List[Wallet]:
        """List wallets.

        Args:
            agent_id: Filter by owner agent ID
            limit: Maximum number of wallets to return
            timeout: Optional request timeout

        Returns:
            List of Wallet objects
        """
        params: Dict[str, Any] = {"limit": limit}
        if agent_id:
            params["agent_id"] = agent_id

        data = self._get("wallets", params=params, timeout=timeout)

        if isinstance(data, list):
            return [Wallet.model_validate(item) for item in data]
        return [Wallet.model_validate(item) for item in data.get("wallets", data.get("items", []))]

    def get_balance(
        self,
        wallet_id: str,
        chain: str = "base",
        token: str = "USDC",
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> WalletBalance:
        """Get wallet balance from chain (non-custodial, read-only).

        Args:
            wallet_id: The wallet ID
            chain: Chain identifier (e.g., "base", "polygon")
            token: The token symbol (e.g., USDC, USDT)
            timeout: Optional request timeout

        Returns:
            WalletBalance object with balance from chain
        """
        params = {"chain": chain, "token": token}
        data = self._get(f"wallets/{wallet_id}/balance", params=params, timeout=timeout)
        return WalletBalance.model_validate(data)

    def get_addresses(
        self,
        wallet_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Dict[str, str]:
        """Get all wallet addresses (chain -> address mapping).

        Args:
            wallet_id: The wallet ID
            timeout: Optional request timeout

        Returns:
            Dictionary mapping chain names to addresses
        """
        return self._get(f"wallets/{wallet_id}/addresses", timeout=timeout)

    def set_address(
        self,
        wallet_id: str,
        chain: str,
        address: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Wallet:
        """Set wallet address for a chain.

        Args:
            wallet_id: The wallet ID
            chain: Chain identifier (e.g., "base", "polygon")
            address: Wallet address on the chain
            timeout: Optional request timeout

        Returns:
            Updated Wallet object
        """
        payload = {"chain": chain, "address": address}
        data = self._post(f"wallets/{wallet_id}/addresses", payload, timeout=timeout)
        return Wallet.model_validate(data)

    def update(
        self,
        wallet_id: str,
        limit_per_tx: Optional[Decimal] = None,
        limit_total: Optional[Decimal] = None,
        is_active: Optional[bool] = None,
        metadata: Optional[Dict[str, Any]] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Wallet:
        """Update a wallet.

        Args:
            wallet_id: The wallet ID
            limit_per_tx: New per-transaction spending limit
            limit_total: New total spending limit
            is_active: Enable/disable the wallet
            metadata: New metadata
            timeout: Optional request timeout

        Returns:
            Updated Wallet object
        """
        payload: Dict[str, Any] = {}
        if limit_per_tx is not None:
            payload["limit_per_tx"] = str(limit_per_tx)
        if limit_total is not None:
            payload["limit_total"] = str(limit_total)
        if is_active is not None:
            payload["is_active"] = is_active
        if metadata is not None:
            payload["metadata"] = metadata

        data = self._patch(f"wallets/{wallet_id}", payload, timeout=timeout)
        return Wallet.model_validate(data)

    def transfer(
        self,
        wallet_id: str,
        *,
        destination: str,
        amount: Decimal,
        token: str = "USDC",
        chain: str = "base_sepolia",
        domain: str = "localhost",
        memo: Optional[str] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> WalletTransferResponse:
        """Transfer stablecoins from a wallet (agent is sender)."""
        payload: Dict[str, Any] = {
            "destination": destination,
            "amount": str(amount),
            "token": token,
            "chain": chain,
            "domain": domain,
        }
        if memo is not None:
            payload["memo"] = memo
        data = self._post(f"wallets/{wallet_id}/transfer", payload, timeout=timeout)
        return WalletTransferResponse.model_validate(data)


__all__ = [
    "AsyncWalletsResource",
    "WalletsResource",
]
